src/nogpu/handle.cpp

```diff
@@ -99,63 +97,131 @@ void Handle::ReadTo(void* /* data */,
+                                           kernel_src);
+    return this->Run(obj);
 }
 
-Invoker Handle::PrepareInvoker(const InvokerFactory& /* factory */,
-                               const std::vector<solver::KernelInfo>& /* kernels */) const
+Invoker Handle::PrepareInvoker(const InvokerFactory& factory,
+                               const std::vector<solver::KernelInfo>& kernels) const
 {
-    return {};
+    std::vector<Kernel> built;
+    for(auto& k : kernels)
+    {
+        MIOPEN_LOG_I2("Preparing kernel: " << k.kernel_name);
+        const auto kernel = this->impl->cache.AddKernel(*this,
+                                                        "",
+                                                        "",
+                                                        k.kernel_file,
+                                                        k.kernel_name,
+                                                        k.l_wk,
+                                                        k.g_wk,
+                                                        k.comp_options,
+                                                        kernels.size());
+        built.push_back(kernel);
+    }
+    return factory(built);
 }
 
-void Handle::ClearKernels(const std::string& /* algorithm */,
-                          const std::string& /* network_config */) const
+void Handle::ClearKernels(const std::string& algorithm, const std::string& network_config) const
 {
+    this->impl->cache.ClearKernels(algorithm, network_config);
 }
 
-const std::vector<Kernel>& Handle::GetKernelsImpl(const std::string& /* algorithm */,
-                                                  const std::string& /* network_config */) const
+const std::vector<Kernel>& Handle::GetKernelsImpl(const std::string& algorithm,
+                                                  const std::string& network_config) const
 {
-    static std::vector<Kernel> tmp;
-    return tmp;
+    return this->impl->cache.GetKernels(algorithm, network_config);
 }
 
-bool Handle::HasKernel(const std::string& /* algorithm */,
-                       const std::string& /* network_config */) const
+bool Handle::HasKernel(const std::string& algorithm, const std::string& network_config) const
 {
-    return false;
+    return this->impl->cache.HasKernels(algorithm, network_config);
 }
 
 KernelInvoke Handle::Run(Kernel /* k */) const { return {}; }
 
-Program Handle::LoadProgram(const std::string& /* program_name */,
-                            std::string /* params */,
-                            bool /* is_kernel_str */,
-                            const std::string& /* kernel_src */) const
+Program Handle::LoadProgram(const std::string& program_name,
+                            std::string params,
+                            bool is_kernel_str,
+                            const std::string& kernel_src) const
 {
-    return {};
+    params += " -mcpu=" + this->GetTargetProperties().Name();
```

Comment:
```C+++
    if(!miopen::EndsWith(program_name, ".mlir-cpp"))
    {
        params += " -mcpu=" + this->GetTargetProperties().Name();
    }
```
Otherwise the binary cache would miss MLIR kernels percompiled with HIPNOGPU backend.

Reply:
Caught accidentally.